moshi/src/main/java/com/squareup/moshi/BufferedSinkJsonWriter.java

```diff
@@ -0,0 +1,441 @@
+    return this;
+  }
+
+  private void writeDeferredName() throws IOException {
+    if (deferredName != null) {
+      beforeName();
+      string(deferredName);
+      deferredName = null;
+    }
+  }
+
+  @Override public JsonWriter value(String value) throws IOException {
+    if (value == null) {
+      return nullValue();
+    }
+    if (promoteNameToValue) {
+      return name(value);
+    }
+    writeDeferredName();
+    beforeValue();
+    string(value);
+    pathIndices[stackSize - 1]++;
+    return this;
+  }
+
+  @Override public JsonWriter nullValue() throws IOException {
+    if (deferredName != null) {
+      if (serializeNulls) {
+        writeDeferredName();
+      } else {
+        deferredName = null;
+        return this; // skip the name and the value
+      }
+    }
+    beforeValue();
+    sink.writeUtf8("null");
+    pathIndices[stackSize - 1]++;
+    return this;
+  }
+
+  @Override public JsonWriter value(boolean value) throws IOException {
+    writeDeferredName();
+    beforeValue();
+    sink.writeUtf8(value ? "true" : "false");
+    pathIndices[stackSize - 1]++;
+    return this;
+  }
+
+  @Override public JsonWriter value(double value) throws IOException {
+    if (Double.isNaN(value) || Double.isInfinite(value)) {
+      throw new IllegalArgumentException("Numeric values must be finite, but was " + value);
+    }
+    if (promoteNameToValue) {
+      return name(Double.toString(value));
+    }
+    writeDeferredName();
+    beforeValue();
+    sink.writeUtf8(Double.toString(value));
+    pathIndices[stackSize - 1]++;
+    return this;
+  }
+
+  @Override public JsonWriter value(long value) throws IOException {
+    if (promoteNameToValue) {
+      return name(Long.toString(value));
+    }
+    writeDeferredName();
+    beforeValue();
+    sink.writeUtf8(Long.toString(value));
+    pathIndices[stackSize - 1]++;
+    return this;
+  }
+
+  @Override public JsonWriter value(Number value) throws IOException {
+    if (value == null) {
+      return nullValue();
+    }
+
+    String string = value.toString();
+    if (!lenient
+        && (string.equals("-Infinity") || string.equals("Infinity") || string.equals("NaN"))) {
+      throw new IllegalArgumentException("Numeric values must be finite, but was " + value);
+    }
+    if (promoteNameToValue) {
+      return name(string);
+    }
+    writeDeferredName();
+    beforeValue();
+    sink.writeUtf8(string);
+    pathIndices[stackSize - 1]++;
+    return this;
+  }
+
+  /**
+   * Ensures all buffered data is written to the underlying {@link Sink}
+   * and flushes that writer.
+   */
+  public void flush() throws IOException {
+    if (stackSize == 0) {
+      throw new IllegalStateException("JsonWriter is closed.");
+    }
+    sink.flush();
+  }
+
+  /**
+   * Flushes and closes this writer and the underlying {@link Sink}.
+   *
+   * @throws JsonDataException if the JSON document is incomplete.
+   */
+  public void close() throws IOException {
```

Comment:
Fixed (for the future):

<img width="1315" alt="screen shot 2016-01-21 at 10 16 48 am" src="https://cloud.githubusercontent.com/assets/66577/12484419/3c8c97de-c028-11e5-8d11-5004f969359e.png">